### ISSUE TITLE
Replaced usage of dicts with nonsense keys with sets in kqueuereactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6684]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#6684|
|Type|enhancement|
|Created|2013-08-19 01:54:44Z|
|Branch|https://github.com/twisted/twisted/tree/kqueue-sets-6684|

This should save memory, and with it a bit of CPU time.

Attachments:

* [sets.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/048/04890cff9d5d33f43f90acca27d940e50f1d5620/671535831ac3e203b1a6d89dd469cbd7d6dfbf9a.diff) (3178 bytes) - added by Alex on 2013-08-19 01:54:59Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6684 6684
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__ 
branch__branches_kqueue_sets_6684 branches/kqueue-sets-6684
branch_author__Alex Alex
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1376877284000000 1376877284000000
changetime__1380936186000000 1380936186000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Alex commented|
|-|-|

A benchmark demonstrating the claim that `sets` will be faster than `dicts`:
```
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = {2: 1}" "s[1] = 1; del s[1]"
100000000 loops, best of 3: 0.0196 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = {2: 1}" "s[1] = 1; del s[1]"
100000000 loops, best of 3: 0.0196 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = {2: 1}" "s[1] = 1; del s[1]"
100000000 loops, best of 3: 0.0196 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$
Alexanders-MacBook-Pro:twisted alex_gaynor$
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = set([2])" "s.add(1); s.remove(1)"
100000000 loops, best of 3: 0.0182 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = set([2])" "s.add(1); s.remove(1)"
100000000 loops, best of 3: 0.0182 usec per loop
Alexanders-MacBook-Pro:twisted alex_gaynor$ pypy -mtimeit -s "s = set([2])" "s.add(1); s.remove(1)"
100000000 loops, best of 3: 0.0183 usec per loop
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

For what it's worth, the microbenchmark here doesn't seem very interesting to me.  Replacing dicts used like sets with sets seems like a fine change to make but a compelling *performance* argument should involve a benchmark of the *reactor*.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [40173]) Branching to 'kqueue-sets-6684'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [40174]) Apply sets.diff 

refs [#6684](https://github.com/twisted/twisted/issues/6684)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun|
|-|-|

As I mentioned above, this seems like a fine change.  I added a misc news file and buildbot thinks the branch is good.  Also as mentioned above, the microbenchmark doesn't factor into my consideration of this change (if it did then the fact that set operations take 2x as long on CPython might influence the outcome here :).  This is just a nice simplification of the implementation.

Thanks for the patch, Alex.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [40176]) Merge kqueue-sets-6684

Author: Alex
Reviewer: exarkun
Fixes: [#6684](https://github.com/twisted/twisted/issues/6684)

Replace file descriptor storage in the kqueue reactor with a more
suitable data structure - sets.
